### PR TITLE
Port CustomMarshalers.dll to C# and .NET Core

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Port the CustomMarshalers.dll from .NET Framework to .NET Core and put the code in System.Private.CoreLib.

Adds the [System.Runtime.InteropServices.CustomMarshalers](https://docs.microsoft.com/en-us/dotnet/api/system.runtime.interopservices.custommarshalers) namespace with the types that are relevant/implementable on .NET Core.

Also adds relevant (Windows-only) tests.

Supercedes dotnet/corefx#33773

### CODE REVIEW

File: src/System.Private.CoreLib/shared/System/Collections/IEnumerable.cs [hunk 1]
```diff
@@ -7,6 +7,7 @@
 
 namespace System.Collections
 {
+    [ComVisible(true)]
```
Comment (jkotas) [inline src/System.Private.CoreLib/shared/System/Collections/IEnumerable.cs]: Why is this needed?

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Text;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    public class EnumerableToDispatchMarshaler : ICustomMarshaler
```
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: Where is it manually used?

I have been working under assumption that these types are just internal implementation details, not referenced by user code: https://github.com/dotnet/corefx/pull/33773#issuecomment-443389274

File: src/vm/mlinfo.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@
 
 
 #ifdef FEATURE_COMINTEROP
-    DEFINE_ASM_QUAL_TYPE_NAME(ENUMERATOR_TO_ENUM_VARIANT_CM_NAME, g_EnumeratorToEnumClassName, "CustomMarshalers", VER_ASSEMBLYVERSION_STR, g_FXKeyToken);
+    DEFINE_ASM_QUAL_TYPE_NAME(ENUMERATOR_TO_ENUM_VARIANT_CM_NAME, g_EnumeratorToEnumClassName, g_MscorlibAsmName, VER_ASSEMBLYVERSION_STR, g_ECMAKeyToken);
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/vm/mlinfo.cpp]: Is there a way to get an assembly qualified name from the definitions in `mscorlib.h`?

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: As of now it's used in the tests only. But it is public surface area and has to be manually used, at least for getting a managed `IEnumerable` from a native `IDispatch*`. That direction currently throws without the marshaler added with a `MarshalAs` attribute.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: So is the plan to also expose these types via CoreFX as well?

File: src/vm/classnames.h [hunk 4]
```diff
@@ -27,6 +27,7 @@
 #ifdef FEATURE_COMINTEROP
 #define g_ECMAKeyToken "B77A5C561934E089"       // The ECMA key used by some framework assemblies: mscorlib, system, etc.
 #define g_FXKeyToken "b03f5f7f11d50a3a"         // The FX key used by other framework assemblies: System.Web, System.Drawing, etc.
+#define g_MscorlibAsmName "System.Private.CoreLib"
```
Comment (jkotas) [inline src/vm/classnames.h]: I would call it just `g_CorelibAsmName`. (It is what we have been trying to do in other places.)

File: src/vm/mlinfo.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/vm/mlinfo.cpp]: Ok, I have looked how this is used. I think what you have here is fine - the assembly name goes too deep.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorToEnumVariantMarshaler.cs [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Runtime.InteropServices.ComTypes;
+using System.Text;
+
+namespace System.Runtime.InteropServices.CustomMarshalers
+{
+    public class EnumeratorToEnumVariantMarshaler : ICustomMarshaler
```
[the remark this answers is not in the document]
Comment (AaronRobinsonMSFT) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorToEnumVariantMarshaler.cs]: I would prefer to keep these internal until we get told they are needed in an external capacity.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumVariantViewOfEnumerator.cs [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+            Enumerator = enumerator;
+        }
+
+        public IEnumerator Enumerator { get; }
+
+        public IEnumVARIANT Clone()
+        {
+            if (Enumerator is ICloneable clonable)
+            {
+                return new EnumVariantViewOfEnumerator(clonable.Clone() as IEnumerator);
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumVariantViewOfEnumerator.cs]: This is the wrong use of `as`. This will return `null` which will then cause the constructor to throw. The real issue is the type is wrong - so use a C-style case and force the cast exception which will indicate the real issue.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorViewOfEnumVariant.cs [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+            {
+                Marshal.ThrowExceptionForHR(hr);
+            }
+        }
+
+        public object GetUnderlyingObject()
+        {
+            return _enumVariantObject;
+        }
+
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorViewOfEnumVariant.cs]: Extra line

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorViewOfEnumVariant.cs [hunk 8]
```diff
@@ -0,0 +1,70 @@
...
+            return true;
+        }
+
+        public void Reset()
+        {
+            int hr = _enumVariantObject.Reset();
+            if (hr < 0)
+            {
+                Marshal.ThrowExceptionForHR(hr);
```
Comment (AaronRobinsonMSFT) [inline src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorViewOfEnumVariant.cs]: Is this a case where the interop defintion of the interface doesn't expose the `HRESULT`?

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h [hunk 9]
```diff
@@ -0,0 +1,222 @@
...
+        return ++refCount;
+    }
+
+    ULONG STDMETHODCALLTYPE Release() override
+    {
+        ULONG newRefCount = --refCount;
+
+        if (newRefCount == 0)
+        {
+            CoreClrFree(this);
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h]: This is an invalid use of placement new and delete. This does not properly clean up (i.e. destructor is not called). It should be called explicitly if the object must be allocated this way. I am not sure it needs to be though since the marshaller should provide an RCW no? Unless `IEnumVARIANT` isn't marshalled as `IUnknown` - I honestly have never used these marshallers.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+
+        return newRefCount;
+    }
+
+    HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void** ppIntf)
+    {
+        HRESULT hr = E_NOINTERFACE;
+        if (iid == __uuidof(IUnknown))
+        {
+            *ppIntf = dynamic_cast<IUnknown*>(this);
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h]: Bah. Please don't ever use `dynamic_cast<>` - especially from the Windows compiler. Use `static_cast<>` instead.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h [hunk 11]
```diff
@@ -0,0 +1,222 @@
...
+        if (newRefCount == 0)
+        {
+            CoreClrFree(this);
+        }
+
+        return newRefCount;
+    }
+
+    HRESULT STDMETHODCALLTYPE QueryInterface(REFIID iid, void** ppIntf)
+    {
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h]: See the COM tests for implementations of `QueryInterface()`. This is more a question of style, but there is a canonical way to implement `QueryInterface()` as to avoid common bugs.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h (hunk 9 above)
Comment (AaronRobinsonMSFT) [reply tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h]: We should chat if this method of allocation is needed. I would like to understand why if it is - performance?

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+        if(FAILED(hr) || numFetched != 1)
+        {
+            return hr;
+        }
+
+        if (V_I4(&element) != i)
+        {
+            return E_UNEXPECTED;
+        }
+    }
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp]: Please call `VariantClear()` after each loop. When copy/paste happens I want the logic to be sound.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorViewOfEnumVariant.cs (hunk 8 above)
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumeratorViewOfEnumVariant.cs]: This is an implementation of the `IEnumerator.Reset` method. The `IEnumVARIANT` interface exposes the `HRESULT`, so I was planning on throwing the exception in managed code.

File: tests/src/Interop/common/ComHelpers.h [hunk 13]
```diff
@@ -17,7 +17,7 @@
 namespace Internal
 {
     template<typename C, typename I>
-    HRESULT __QueryInterfaceImpl(
+    typename std::enable_if<std::is_assignable<I, C>::value, HRESULT>::type __QueryInterfaceImpl(
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/common/ComHelpers.h]: Why is this needed?

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp [hunk 14]
```diff
@@ -0,0 +1,103 @@
...
+#include <xplatform.h>
+
+extern "C" DLL_EXPORT HRESULT STDMETHODCALLTYPE GetIntegerEnumerator(int start, int count, IEnumVARIANT** ppEnum)
+{
+    if (count < 0)
+    {
+        return E_INVALIDARG;
+    }
+
+    *ppEnum = new(CoreClrAlloc(sizeof(IntegerEnumerator))) IntegerEnumerator(start, count);
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp]: Placement new is no longer valid and will more than likely crash now. Same below.

File: tests/src/Interop/common/ComHelpers.h (hunk 13 above)
Comment (jkoritzinsky) [reply tests/src/Interop/common/ComHelpers.h]: I thought it would be useful, but after thinking over it again, the compiler will likely already do what this was checking for me.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: > So is the plan to also expose these types via reference assemblies in CoreFX as well?

Ping ... I would be curious to understand the plan here.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+{
+    static class IEnumeratorNative
+    {
+        [DllImport(nameof(IEnumeratorNative), PreserveSig = false)]
+        public static extern IEnumerator GetIntegerEnumerator(
+            int start,
+            int count);
+
+        [DllImport(nameof(IEnumeratorNative), PreserveSig = false)]
+        [return: MarshalAs(UnmanagedType.CustomMarshaler, MarshalType = "System.Runtime.InteropServices.CustomMarshalers.EnumerableToDispatchMarshaler, System.Private.CoreLib, Culture=neutral, PublicKeyToken=B77A5C561934E089, Version=4.0.0.0")]
```
Comment (jkotas) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs]: Given the other conversation, I assume that we expect customer code to do this as well. This should be changed to use the type from the public location. It would  be useful to add TODO/comment about it.

File: tests/src/Interop/common/ComHelpers.h (hunk 13 above)
Comment (AaronRobinsonMSFT) [reply tests/src/Interop/common/ComHelpers.h]: Well basically if `I` isn't assignable from `C` this entire QI implementation would be wrong. It has to be true. Yes, please remove.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs (hunk 15 above)
Comment (jkoritzinsky) [reply tests/src/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs]: Until we expose these types in corefx, that is correct. However, as we talked about before, we didn't find any customers (or a number so small that it showed up as 0% on apiport) directly using this marshaler. Therefore, I don't think it's a pressing issue to expose these marshalers via corefx. The only reason I'm doing what I'm doing here is so that we can get some tests for them.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: My initial thought would be to expose them via a CustomMarshalers.dll ref assembly in corefx. However, since they're so rarely used, I could see us just putting them into the System.Runtime.InteropServices ref assembly. For now I'm thinking we just keep them internal and have the tests for them until we get customers asking for it, at which point we would expose it.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs (hunk 15 above)
Comment (jkotas) [reply tests/src/Interop/PInvoke/IEnumerator/IEnumeratorTest.cs]: ok

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: I assume that all these marshaler gets used even without directly referencing the type. Is that correct?

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: The other one does. This one does not. This one requires a direct (or indirect as I'm currently doing in the tests) reference. I'd prefer to keep both in this PR for completeness.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: Do we need to have it build into the product then? It looks like half-done feature. We should either not have it in the product at all; or we should have it in the product and be exposed as a proper public API. I do not think we want to be ever telling people to reference System.Private.CoreLib directly.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: I think this one used to be required in .NET Framework but its functionality was later integrated into the runtime. The fact that this marshaller isn't needed on the parameter but is required on the return value (which gives a "System.PlatformNotSupportedException: IDispatch and IDispatchEx are not supported" error when enumerating without it) is what prompts that idea in my mind.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: I'm fine exposing it via CoreFX in a ref assembly if that's what we want to do. Would you prefer putting it in a CustomMarshalers.dll (matching .NET Framework), or in System.Runtime.InteropServices (to avoid creating another ref-assembly).

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp [hunk 16]
```diff
@@ -0,0 +1,108 @@
...
+    {
+        return hr;
+    }
+
+    if(!((V_VT(&result) == VT_UNKNOWN) || (V_VT(&result) == VT_DISPATCH)))
+    {
+        return E_UNEXPECTED;
+    }
+
+    IEnumVARIANT* pEnum;
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp]: Who is calling release on this?

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: The question for me is whether we need to have it all.

Is all Office interop going to work when we do not have it all (ie not even in CoreLib)? If the answer to this is yet, I would not have it at all.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h [hunk 17]
```diff
@@ -0,0 +1,153 @@
...
+        WORD wFlags,
+        DISPPARAMS *pDispParams,
+        VARIANT *pVarResult,
+        EXCEPINFO *pExcepInfo,
+        UINT *puArgErr) override
+    {
+        if (dispIdMember == DISPID_NEWENUM && (wFlags & INVOKE_PROPERTYGET) == INVOKE_PROPERTYGET)
+        {
+            V_VT(pVarResult) = VT_UNKNOWN;
+            V_UNKNOWN(pVarResult) = new (CoreClrAlloc(sizeof(IntegerEnumerator))) IntegerEnumerator(start, count);
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h]: Invalid placement new.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: I'm not sure. By the way, I found the portion of the code that is causing that error I mentioned above.

https://github.com/dotnet/coreclr/blob/71255bc05db24f078f334ead245898b817c8a06e/src/vm/mngstdinterfaces.cpp#L993-L995

If we implement this, then we don't need `EnumerableToDispatchMarshaler` at all (its functionality will be completely implemented within the runtime).

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h [hunk 18]
```diff
@@ -0,0 +1,153 @@
...
+
+    HRESULT STDMETHODCALLTYPE Reset(void) override
+    {
+        current = start;
+        return S_OK;
+    }
+
+    HRESULT STDMETHODCALLTYPE Clone(IEnumVARIANT **ppEnum) override
+    {
+        IntegerEnumerator* clone = new(CoreClrAlloc(sizeof(IntegerEnumerator))) IntegerEnumerator(start, count);
```
Comment (AaronRobinsonMSFT) [inline tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.h]: Invalid placement new.

File: tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp (hunk 16 above)
Comment (jkoritzinsky) [reply tests/src/Interop/PInvoke/IEnumerator/IEnumeratorNative.cpp]: I should be doing that.... Adding that now.

**Comment by AaronRobinsonMSFT** [comment]
There are a bunch of duplicate issues that I have already commented on. I am not going to comment on each instance. Please be proactive and consider all my style comments and usage issues across the entire change.

**Comment by jkoritzinsky** [comment]
Sorry about that. I'm trying to fix them all but am not being as thorough as I should be. I'll be sure to be more careful.

File: src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: By looking at the comment a little higher up in the file, it seems like the runtime should be using this marshaler automatically. I'll investigate a little more to see what we should be doing.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: > If we implement this, then we don't need EnumerableToDispatchMarshaler at all (its functionality will be completely implemented within the runtime).

That would be preferable, I think.

[the remark this answers is not in the document]
Comment (jkotas) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: It should be just about reverting https://github.com/dotnet/coreclr/pull/11865/ . That fix was made to avoid confusing error message - that was caused by these marshalers missing.

[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/src/System/Runtime/InteropServices/CustomMarshalers/EnumerableToDispatchMarshaler.cs]: That change is now in this PR. Now we don't need to expose this API to the user for it to be used correctly.

File: src/System.Private.CoreLib/ILLinkTrim.xml [hunk 19]
```diff
@@ -40,6 +40,7 @@
     <!-- Accessed via reflection in TraceLogging-style EventSource events. -->
     <type fullname="*f__AnonymousType*" />
     <!-- Accessed via native code. -->
+    <type fullname="System.Runtime.InteropServices.ComTypes.IEnumerable" />
```
Comment (jkotas) [inline src/System.Private.CoreLib/ILLinkTrim.xml]: Do you really need it here? There are other similar interfaces, like `IExpando`, that are rooted implicitly by other means.

**Comment by jkoritzinsky** [comment]
@dotnet-bot test Windows_NT x64 full_opt ryujit CoreCLR Perf Tests Correctness 
@dotnet-bot test  Windows_NT x64 min_opt ryujit CoreCLR Perf Tests Correctness 
@dotnet-bot test Windows_NT x86 full_opt ryujit CoreCLR Perf Tests Correctness 
@dotnet-bot test  Windows_NT x86 min_opt ryujit CoreCLR Perf Tests Correctness

File: src/System.Private.CoreLib/ILLinkTrim.xml (hunk 19 above)
Comment (jkoritzinsky) [reply src/System.Private.CoreLib/ILLinkTrim.xml]: This type needs to be rooted somewhere otherwise it gets linked out. I'm currently looking to see if there's a better place to root it.

Related: we should extend the `CreateRuntimeRootILLinkDescriptorFile` task to also parse the src/vm/mngstditflist.h file since that's where this type (along with the types for other custom marshalers that we haven't added since their types aren't exposed from System.Private.CoreLib) are rooted.

Comment (jkotas) [reply src/System.Private.CoreLib/ILLinkTrim.xml]: Ok. 

> we should extend the CreateRuntimeRootILLinkDescriptorFile task

It is probably not worth the added complexity. The list in `mngstditflist.h` is very unlikely to change, so optimizing for changing it is not worth it.

Comment (jkoritzinsky) [reply src/System.Private.CoreLib/ILLinkTrim.xml]: So I could add these types to `mscorlib.h`, but they'd only be used by the `CreateRuntimeRootILLinkDescriptorFile` task to ensure they aren't linked away. The definitions in `mscorlib.h` would never be used by the runtime with the current code.

Comment (jkotas) [reply src/System.Private.CoreLib/ILLinkTrim.xml]: I think the way you have it right now is fine.

File: src/vm/mngstdinterfaces.cpp [hunk 20]
```diff
@@ -990,9 +990,7 @@ FCIMPL1(Object*, StdMngIEnumerable::GetEnumerator, Object* refThisUNSAFE)
 
     if (retVal == NULL)
     {
-        // In desktop CLR we'll attempt to call through IDispatch(DISPID_NEWENUM)
-        // This is not supported in CoreCLR
-        COMPlusThrow(kPlatformNotSupportedException, IDS_EE_ERROR_IDISPATCH);
```
Comment (AaronRobinsonMSFT) [inline src/vm/mngstdinterfaces.cpp]: Please make sure to remove `IDS_EE_ERROR_IDISPATCH` if it isn't used anywhere else.

**Comment by jjxtra** [comment]
Is this going to make it into .net core 2.x?

**Comment by jkotas** [comment]
The change is part of larger set of changes for full COM support in .NET Core. We do not plan to port  the full COM support back to 2.x.
